src/coreclr/src/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs

```diff
@@ -97,63 +100,116 @@ public static DependencyList ScanAndProcessReturnValue(NodeFactory factory, Flow
+            foreach (var namedArgument in arguments.NamedArguments)
             {
-                var reflectionContext = new ReflectionPatternContext(_context, true, field.DeclaringType.Methods[0], field);
-                reflectionContext.AnalyzingPattern();
-                RequireDynamicallyAccessedMembers(ref reflectionContext, annotation, valueNode, field);
+                TypeSystemEntity entity = null;
+                DynamicallyAccessedMemberTypes annotation = DynamicallyAccessedMemberTypes.None;
+                Origin targetContext = null;
+                if (namedArgument.Kind == CustomAttributeNamedArgumentKind.Field)
+                {
+                    FieldDesc field = attributeType.GetField(namedArgument.Name);
+                    if (field != null)
+                    {
+                        annotation = flowAnnotations.GetFieldAnnotation(field);
+                        entity = field;
+                        targetContext = new FieldOrigin(field);
+                    }
+                }
+                else
+                {
+                    Debug.Assert(namedArgument.Kind == CustomAttributeNamedArgumentKind.Property);
+                    PropertyPseudoDesc property = ((MetadataType)attributeType).GetProperty(namedArgument.Name, null);
+                    MethodDesc setter = property.SetMethod;
+                    if (setter != null && setter.Signature.Length > 0 && !setter.Signature.IsStatic)
+                    {
+                        annotation = flowAnnotations.GetParameterAnnotation(setter, 1);
+                        entity = property;
+                        targetContext = new ParameterOrigin(setter, 1);
+                    }
+                }
+
+                if (annotation != DynamicallyAccessedMemberTypes.None)
+                {
+                    ValueNode valueNode = GetValueNodeForCustomAttributeArgument(namedArgument.Value);
+                    if (valueNode != null)
+                    {
+                        var reflectionContext = new ReflectionPatternContext(logger, true, method, targetContext);
+                        try
+                        {
+                            reflectionContext.AnalyzingPattern();
+                            var scanner = new ReflectionMethodBodyScanner(factory, flowAnnotations, logger);
+                            scanner.RequireDynamicallyAccessedMembers(ref reflectionContext, annotation, valueNode, targetContext);
+                            if (result == null)
+                            {
+                                result = scanner._dependencies;
+                            }
+                            else
+                            {
+                                result.AddRange(scanner._dependencies);
+                            }
+                        }
+                        finally
+                        {
+                            reflectionContext.Dispose();
```

Comment:
This makes it look like that `ReflectionPatternContext` has some resources that have to be reclaimed, but it is actually just a debug check.

Instead of finally block and Dispose, would it be better to have to debug-only method to do the debug-only check?

Reply:
(Ignore this if this is just mirroring ILLinker structure.)

Reply:
Yeah, it's from IL linker. The `ReflectionPatternContext` pre-dates the new dataflow analysis (it was introduced for the peephole scanner). It deals with the fact that one of the places that uses it has many exits from the method and it's a chore to manually call it in each of those spots.

`IDisposable` sounds good in theory (could be used with `using`), but then we have to pass the thing as a `ref` and C# forbids passing things we're `using` like that, leaving us with manual `finally`.